04-onion-routing.md

```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Route Blinding
+
+Nodes receiving onion packets may hide their identity from senders by
+"blinding" an arbitrary amount of hops at the end of an onion path.
+
+When using route blinding, nodes find a route to themselves from a given
+"introduction node". They then use ECDH with each node in that route to create
+a "blinded" node ID and an encrypted blob (`encrypted_data`) for each one of
+the blinded nodes.
+
+They communicate this blinded route and the encrypted blobs to the sender.
+The sender finds a route to the introduction node and extends it with the
+blinded route provided by the recipient. The sender includes the encrypted
+blobs in the corresponding onion payloads: they allow nodes in the blinded
+part of the route to "unblind" the next node and correctly forward the packet.
+
+The `encrypted_data` is a TLV stream, encrypted for a given blinded node, that
+may contain the following TLV fields:
+
+1. `tlv_stream`: `encrypted_data_tlv`
+2. types:
+    1. type: 1 (`padding`)
+    2. data:
+        * [`...*byte`:`padding`]
+    1. type: 2 (`short_channel_id`)
+    2. data:
+        * [`short_channel_id`:`short_channel_id`]
+    1. type: 4 (`next_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+    1. type: 6 (`path_id`)
+    2. data:
+        * [`...*byte`:`data`]
+    1. type: 8 (`next_blinding_override`)
+    2. data:
+        * [`point`:`blinding`]
+
+#### Requirements
+
+A recipient N(r) creating a blinded route `N(0) -> N(1) -> ... -> N(r)` to itself:
+
+- MUST create a blinded node ID `B(i)` for each node using the following algorithm:
+  - `e(0) <- {0;1}^256`
+  - `E(0) = e(0) * G`
+  - For every node `N(i)` in the route:
+    - let `P(i) = k(i) * G` be `N(i)`'s `node_id` (`k(i)` is `N(i)`'s private key)
+    - `ss(i) = SHA256(e(i) * P(i)) = SHA256(k(i) * E(i))` (ECDH shared secret known only by `N(r)` and `N(i)`)
+    - `B(i) = HMAC256("blinded_node_id", ss(i)) * P(i)` (blinded `node_id` for `N(i)`, private key known only by `N(i)`)
+    - `rho(i) = HMAC256("rho", ss(i))`
+    - `e(i+1) = SHA256(E(i) || ss(i)) * e(i)` (ephemeral private key, only known by `N(r)`)
+    - `E(i+1) = SHA256(E(i) || ss(i)) * E(i)` (NB: `N(i)` MUST NOT learn `e(i)`)
```

Comment:
Since we made the ECDH explicit in line 430 we could also here help a bit by writing:

```
E(i+1) = e(i+1)*G 
       = SHA256(E(i) || ss(i)) * e(i) * G
       = SHA256(E(i) || ss(i)) * E(i)
```

with the comment from line 427 everyone can see that the math is correctly